### PR TITLE
osd: improve error message when FileStore op fails due to EPERM

### PR DESCRIPTION
**Pull request by smithfarm**
References: http://tracker.ceph.com/issues/18037
Signed-off-by: Nathan Cutler <ncutler@suse.com>

### CODE REVIEW

**Comment by smithfarm** [comment]
[[Bug 1010279] L3: OSD fails to start - leveldb corruption leads to "Operation not permitted not handled" and assert](https://trello.com/c/r42lLZB6/2209-bug-1010279-l3-osd-fails-to-start-leveldb-corruption-leads-to-operation-not-permitted-not-handled-and-assert)

File: src/os/filestore/FileStore.cc [hunk 1]
```diff
@@ -2997,6 +2997,10 @@ void FileStore::_do_transaction(
 	  msg = "ENOTEMPTY suggests garbage data in osd data dir";
 	}
 
+        if (r == -EPERM) {
```
Comment (jecluis) [inline src/os/filestore/FileStore.cc]: personally, i'd go with an `else if` branching off after the previous if. otherwise it looks good to me.

**Comment by athanatos** [comment]
This generally looks ok to me (I support @jecluis 's comment though).

**Comment by smithfarm** [comment]
@jecluis @athanatos Please have another look. I got access to the cluster where the EPERM was manifesting and was very surprised to find that "chown -R ceph.ceph /var/lib/ceph/osd/..." did not fix it. The leveldb was corrupted, though, and repairing it made the error go away (though the OSD still refused to start, but for a different reason).

Based on the above, I revised the message and also converted all the if clauses into else-ifs hanging off the first if in the series.

**Comment by smithfarm** [comment]
@jecluis Could you re-review?

File: src/os/filestore/FileStore.cc [hunk 2]
```diff
@@ -2994,15 +2994,16 @@ void FileStore::_do_transaction(
...
-	if (r == -ENOSPC)
+	else if (r == -ENOSPC)
 	  // For now, if we hit _any_ ENOSPC, crash, before we do any damage
 	  // by partially applying transactions.
 	  msg = "ENOSPC handling not implemented";
-
-	if (r == -ENOTEMPTY) {
+	else if (r == -ENOTEMPTY) {
 	  msg = "ENOTEMPTY suggests garbage data in osd data dir";
 	}
```
Comment (athanatos) [inline src/os/filestore/FileStore.cc]: } else if {

File: src/os/filestore/FileStore.cc [hunk 3]
```diff
@@ -2994,15 +2994,16 @@ void FileStore::_do_transaction(
 			     op->op == Transaction::OP_CLONE ||
 			     op->op == Transaction::OP_CLONERANGE2))
 	  msg = "ENOENT on clone suggests osd bug";
-
-	if (r == -ENOSPC)
+	else if (r == -ENOSPC)
```
Comment (athanatos) [inline src/os/filestore/FileStore.cc]: add braces around the first if and make it all consistent.

**Comment by smithfarm** [comment]
@athanatos Done!

**Comment by athanatos** [comment]
Just needs to be tested.

**Comment by smithfarm** [comment]
Pushed `wip-18037` to shaman.

`teuthology-suite --priority 101 --suite rados --subset $(expr $RANDOM % 2000)/2000 --email ncutler@suse.cz --ceph wip-18037 --machine-type smithi -k distro`

**pending build**

**Comment by smithfarm** [comment]
@athanatos Is the above run sufficient?

**Comment by athanatos** [comment]
Assuming it was a rados suite run, sure.

**Pull request merged**